scipy/spatial/tests/test_distance.py

```diff
@@ -2215,6 +2209,51 @@ def test_immutable_input(metric):
         getattr(scipy.spatial.distance, metric)(x, x, w=x)
 
 
+class TestDegenerateInput:
+    """Degenerate input to boolean metrics should give zero similarity.
+    See gh-21135.
+    """
+
+    # kulczynski1 and sokalmichener are not covered because they are to be
+    # deprecated and removed.
+    @pytest.fixture(scope='class', params=[
+        'dice', 'hamming', 'jaccard', 'rogerstanimoto', 'russellrao',
+        'sokalsneath', 'yule',
+    ])
+    def metric(self, request):
+        return request.param
+
+    def test_empty_input_unweighted_py(self, metric: str):
+        func = getattr(scipy.spatial.distance, metric)
+        u, v = [], []
+        assert_equal(func(u, v), 0.0)
+
+    def test_empty_input_unweighted_xdist(self, metric: str):
+        u, v = [], []
+        assert_equal(pdist([u, v], metric), [0.0])
+        assert_equal(cdist([u], [v], metric), [[0.0]])
+
+    def test_empty_input_weighted_py(self, metric: str):
+        func = getattr(scipy.spatial.distance, metric)
+        u, v, w = [], [], []
+        assert_equal(func(u, v, w), 0.0)
+
+    def test_empty_input_weighted_xdist(self, metric: str):
+        u, v, w = [], [], []
+        assert_equal(pdist([u, v], metric, w=w), [0.0])
+        assert_equal(cdist([u], [v], metric, w=w), [[0.0]])
+
+    def test_zero_weighted_input_py(self, metric: str):
+        func = getattr(scipy.spatial.distance, metric)
+        u, v, w = [1, 0, 1], [0, 1, 0], [0, 0, 0]
+        assert_equal(func(u, v, w), 0.0)
+
+    def test_zero_weighted_input_xdist(self, metric: str):
+        u, v, w = [1, 0, 1], [0, 1, 0], [0, 0, 0]
+        assert_equal(pdist([u, v], metric, w=w), [0.0])
+        assert_equal(cdist([u], [v], metric, w=w), [[0.0]])
```

Comment:
> the handling of degenerate input (such as empty input, zero input, or zero weights) for the Boolean metrics in scipy.spatial.distance varies across metrics / Python vs C++ / weighted vs unweighted.

Taking the case of `russellrao` as an example, this test currently produces a distance of `0.0` when all weights are zero as above.

If I zero out the input vectors as well, the test continues to pass:

```diff
--- a/scipy/spatial/tests/test_distance.py
+++ b/scipy/spatial/tests/test_distance.py
@@ -2249,7 +2249,7 @@ class TestDegenerateInput:
         assert_equal(func(u, v, w), 0.0)
 
     def test_zero_weighted_input_xdist(self, metric: str):
-        u, v, w = [1, 0, 1], [0, 1, 0], [0, 0, 0]
+        u, v, w = [0, 0, 0], [0, 0, 0], [0, 0, 0]
         assert_equal(pdist([u, v], metric, w=w), [0.0])
         assert_equal(cdist([u], [v], metric, w=w), [[0.0]])
```

but, if I keep the zero inputs and provide non-zero weights the distance becomes `1.0` and the test fails for `russellrao` only:

```diff
--- a/scipy/spatial/tests/test_distance.py
+++ b/scipy/spatial/tests/test_distance.py
@@ -2249,7 +2249,7 @@ class TestDegenerateInput:
         assert_equal(func(u, v, w), 0.0)
 
     def test_zero_weighted_input_xdist(self, metric: str):
-        u, v, w = [1, 0, 1], [0, 1, 0], [0, 0, 0]
+        u, v, w = [0, 0, 0], [0, 0, 0], [0.3, 0.6, 0.1]
         assert_equal(pdist([u, v], metric, w=w), [0.0])
         assert_equal(cdist([u], [v], metric, w=w), [[0.0]])
```

Is there clear guidance on the indeterminate case for `russellrao` in the papers? I do see some "decisive" guidance for a few metrics, but not that one?

Reply:
The docs for `russellrao` clearly define the weights as what I'd intuitively expect, `The weights for each value in u and v.`

Reply:
Good catch! Your test case reveals a peculiar property of `russellrao` — if a vector `x` contains any zero, its self-dissimilarity (i.e. the distance between `x` and `x`) is *not* zero. It is actually not specific to the weights; if you set the weights to `[1, 1, 1]` or leave it out, the self-distance is still non-zero.

This peculiar property was pointed out in e.g. [ref]. (To be filled) There the authors noted that the russelrao dissimilarity is not a metric (because a metric must give zero self-distance), but that *if* self-distance is overridden to be zero, then the (modified) dissimilarity actually satisfies the triangle inequality and is qualified as a metric.

In a follow-up PR (gh-21580) I write out the formulas of `russellrao` (among others) for both the weighted and unweighted cases; see [here](https://output.circle-artifacts.com/output/job/13d8741e-c400-4a32-99e1-9aadff3ba7b4/artifacts/0/html/reference/generated/scipy.spatial.distance.russellrao.html) for the rendered output. In particular, I gave an example in the “Examples” section to show that the self dissimilarity is non-zero if the vector contains any zero.